pkg/sources/github/github.go

```diff
@@ -1126,29 +1052,49 @@ func (s *Source) setProgressCompleteWithRepo(index int, offset int, repoURL stri
 	s.SetProgressComplete(index+offset, len(s.repos)+offset, fmt.Sprintf("Repo: %s", repoURL), encodedResumeInfo)
 }
 
-const initialPage = 1 // page to start listing from
-
-func (s *Source) scanComments(ctx context.Context, repoPath string, chunksChan chan *sources.Chunk) error {
-	// Support ssh and https URLs
+func (s *Source) scanComments(ctx context.Context, repoPath string, repoInfo *repoInfo, chunksChan chan *sources.Chunk) error {
+	// Support ssh and https URLs.
 	repoURL, err := git.GitURLParse(repoPath)
 	if err != nil {
 		return err
 	}
+	urlString := repoURL.String()
 
-	trimmedURL := removeURLAndSplit(repoURL.String())
-	if repoURL.Host == "gist.github.com" && s.includeGistComments {
-		return s.processGistComments(ctx, repoPath, trimmedURL, repoURL, chunksChan)
+	urlParts := trimURLAndSplit(urlString)
+	if len(urlParts) < 2 || len(urlParts) > 3 {
+		return fmt.Errorf("invalid repository or gist URL (%s): length of URL segments should be 2 or 3", urlString)
 	}
-	return s.processRepoComments(ctx, repoPath, trimmedURL, repoURL, chunksChan)
+
+	if s.includeGistComments && urlParts[0] == "gist.github.com" {
+		return s.processGistComments(ctx, urlString, urlParts, repoInfo, chunksChan)
+	} else if s.includeIssueComments || s.includePRComments {
+		return s.processRepoComments(ctx, repoInfo, chunksChan)
+	}
+	return nil
 }
 
-func (s *Source) processGistComments(ctx context.Context, repoPath string, trimmedURL []string, repoURL *url.URL, chunksChan chan *sources.Chunk) error {
-	ctx.Logger().V(2).Info("scanning github gist comments", "repository", repoPath)
+// trimURLAndSplit removes extraneous information from the |url| and splits it into segments.
+// This is typically 3 segments: host, owner, and name/ID; however, Gists have some edge cases.
+//
+// Examples:
+// - "https://github.com/trufflesecurity/trufflehog" => ["github.com", "trufflesecurity", "trufflehog"]
+// - "https://gist.github.com/nat/5fdbb7f945d121f197fb074578e53948" => ["gist.github.com", "nat", "5fdbb7f945d121f197fb074578e53948"]
+// - "https://gist.github.com/ff0e5e8dc8ec22f7a25ddfc3492d3451.git" => ["gist.github.com", "ff0e5e8dc8ec22f7a25ddfc3492d3451"]
+func trimURLAndSplit(url string) []string {
+	trimmedURL := strings.TrimPrefix(url, "https://")
```

Comment:
Do you think it's worth it to use `url.Parse` here? For example if the input is `ssh://github.com/...`?

Reply:
+1

Reply:
This method isn't used anywhere else, so it could be inlined in `scanComments` which already calls `url.Parse` via `git.GitURLParse`.

https://github.com/trufflesecurity/trufflehog/blob/7620906b075a7ec53926b4e5eafee4b3ff9c2c13/pkg/sources/github/github.go#L1132-L1133

https://github.com/trufflesecurity/trufflehog/blob/7620906b075a7ec53926b4e5eafee4b3ff9c2c13/pkg/sources/git/git.go#L365-L378

I have added some tests and updated the code a bit to handle SSH, though.